### PR TITLE
Use filedescriptor for workspace and at functions

### PR DESCRIPTION
**Pull request by bapt**

### CODE REVIEW

File: include/error.h [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #define KP_EENCRYPT         8
 #define KP_INVALID_MSG      9
 #define KP_EXIT             10
+#define KP_ENOENT           11
```
Comment (paulfariello) [inline include/error.h]: If you want to report an error with an existing code you should just set
```
errno = ENOENT
ret = KP_ERRNO
```

**Comment by bapt** [comment]
also brings a new feature which concens tell the user a workspace is non exitent and that they should run kickpass init.

The main goal of this PR is to simplify code a bit before, adding more feautres/bugfixes

File: lib/config.c [hunk 2]
```diff
@@ -61,15 +61,11 @@ kp_error_t
 kp_cfg_create(struct kp_ctx *ctx, const char *sub)
 {
 	kp_error_t ret;
-	char path[PATH_MAX] = "";
+	char path[PATH_MAX];
```
Comment (paulfariello) [inline lib/config.c]: IIRC initializing path avoid valgrind complaining about uninitialized variable usage.

File: lib/config.c [hunk 3]
```diff
@@ -61,15 +61,11 @@ kp_error_t
...
 	struct kp_safe cfg_safe;
 
-	if (strlcpy(path , sub, PATH_MAX) >= PATH_MAX) {
-		errno = ENOMEM;
-		return KP_ERRNO;
-	}
-
-	if (strlcat(path , "/" KP_CONFIG_SAFE_NAME, PATH_MAX) >= PATH_MAX) {
+	if (snprintf(path, PATH_MAX, "%s%s" KP_CONFIG_SAFE_NAME,
+	    *sub == '\0' ? "": "/", sub) >= PATH_MAX) {
```
Comment (paulfariello) [inline lib/config.c]: I would prefer a `strlen(sub) == 0` which is a bit more explicit.

File: include/safe.h [hunk 4]
```diff
@@ -49,7 +49,6 @@ kp_error_t kp_safe_load(struct kp_ctx *, struct kp_safe *, const char *);
 kp_error_t kp_safe_save(struct kp_ctx *, struct kp_safe *);
 kp_error_t kp_safe_open(struct kp_ctx *, struct kp_safe *, bool);
 kp_error_t kp_safe_close(struct kp_ctx *, struct kp_safe *);
-kp_error_t kp_safe_get_path(struct kp_ctx *, struct kp_safe *, char *, size_t);
```
Comment (paulfariello) [inline include/safe.h]: I'm not sure we want to get rid of this method.

File: lib/config.c (hunk 3 above)
Comment (paulfariello) [reply lib/config.c]: Btw cool to try to avoid the dirty double / on safe path.

File: lib/config.c [hunk 5]
```diff
@@ -179,25 +171,12 @@ kp_cfg_find(struct kp_ctx *ctx, const char *path, char *cfg_path, size_t size)
...
-		}
-
-		if (strlcat(abspath, "/", PATH_MAX) >= PATH_MAX) {
-			errno = ENOMEM;
-			return KP_ERRNO;
-		}
-
-		if (strlcat(abspath, cfg_path, PATH_MAX) >= PATH_MAX) {
+		if (snprintf(abspath, PATH_MAX, "%s/%s/" KP_CONFIG_SAFE_NAME,
+                    ctx->ws_path, cfg_path) >= PATH_MAX) {
```
Comment (paulfariello) [inline lib/config.c]: a bit simpler :+1:

File: lib/safe.c [hunk 6]
```diff
@@ -53,15 +52,7 @@ kp_safe_load(struct kp_ctx *ctx, struct kp_safe *safe, const char *name)
...
-	if ((ret = kp_safe_get_path(ctx, safe, path, PATH_MAX)) != KP_SUCCESS) {
-		return ret;
-	}
-
-	if (stat(path, &stats) != 0) {
-		return KP_ERRNO;
-	}
-
-	safe->cipher = open(path, O_RDWR | O_NONBLOCK);
+	safe->cipher = openat(ctx->ws_fd, name, O_RDWR | O_NONBLOCK);
```
Comment (paulfariello) [inline lib/safe.c]: Pretty cool! :+1:

File: lib/safe.c [hunk 7]
```diff
@@ -116,22 +110,12 @@ kp_safe_create(struct kp_ctx *ctx, struct kp_safe *safe, const char *name)
 		return ret;
 	}
 
-	if ((ret = kp_safe_get_path(ctx, safe, path, PATH_MAX)) != KP_SUCCESS) {
-		return ret;
-	}
-
-	if ((ret = kp_safe_mkdir(ctx, path)) != KP_SUCCESS) {
+	if ((ret = kp_safe_mkdir(ctx, name)) != KP_SUCCESS) {
```
Comment (paulfariello) [inline lib/safe.c]: I'd rather use `safe->name` in case `kp_safe_init` would do some sanitization on `name`.

File: lib/safe.c [hunk 8]
```diff
@@ -70,21 +61,25 @@ kp_safe_load(struct kp_ctx *ctx, struct kp_safe *safe, const char *name)
...
+	char   path[PATH_MAX], *rdir;
 
 	assert(ctx);
 	assert(path);
 
-	rdir = path + strlen(ctx->ws_path);
-	rdir++; /* Skip first / as it is part of the ctx->ws */
+	if (strlcpy(path, name, PATH_MAX) >= PATH_MAX) {
+		errno = ENOMEM;
+		return (KP_ERRNO);
```
Comment (paulfariello) [inline lib/safe.c]: Why `()`? Better stick to coding style.

File: lib/safe.c [hunk 9]
```diff
@@ -116,22 +110,12 @@ kp_safe_create(struct kp_ctx *ctx, struct kp_safe *safe, const char *name)
...
-	if (stat(path, &stats) == 0) {
-		errno = EEXIST;
-		return KP_ERRNO;
-	} else if (errno != ENOENT) {
-		return errno;
-	}
-
-	safe->cipher = open(path, O_RDWR | O_NONBLOCK | O_CREAT, S_IRUSR | S_IWUSR);
+	safe->cipher = openat(ctx->ws_fd, name,
+	    O_RDWR | O_NONBLOCK | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
```
Comment (paulfariello) [inline lib/safe.c]: :+1:

File: lib/safe.c [hunk 10]
```diff
@@ -354,31 +321,17 @@ kp_error_t
...
 
 	assert(ctx);
 	assert(safe);
 	assert(name);
 
-	if ((ret = kp_safe_get_path(ctx, safe, oldpath, PATH_MAX)) != KP_SUCCESS) {
-		return ret;
-	}
-
-	if (strlcpy(safe->name, name, PATH_MAX) >= PATH_MAX) {
```
Comment (paulfariello) [inline lib/safe.c]: We should keep the API behavior of returning safe with new name after call to `kp_safe_rename`.

File: include/safe.h (hunk 4 above)
Comment (paulfariello) [reply include/safe.h]: After a full review, I'm fine with removing this function.

File: lib/safe.c [hunk 11]
```diff
@@ -464,9 +392,9 @@ kp_safe_store(struct kp_ctx *ctx, struct kp_safe *safe, int timeout)
 	}
 
 	unsafe.timeout = timeout;
-	if ((ret = kp_safe_get_path(ctx, safe, unsafe.path,
-	                            PATH_MAX)) != KP_SUCCESS) {
-		return ret;
+	if (strlcpy(unsafe.path, safe->name, PATH_MAX) >= PATH_MAX) {
```
Comment (paulfariello) [inline lib/safe.c]: We should rename `unsafe.path` into `unsafe.name`.

File: src/main.c [hunk 12]
```diff
@@ -238,11 +236,27 @@ find_command(const char *command)
...
 {
+	extern char *__progname;
 	struct kp_cmd *cmd;
+	kp_error_t ret;
 
 	if (optind >= argc)
 		kp_errx(KP_EINPUT, "missing command");
 
+	ret = kp_init(ctx);
+	if (ret == KP_ENOENT) {
```
Comment (paulfariello) [inline src/main.c]: Maybe we could return an error specific to missing workspace `KP_ENOWORKSPACE` or just use the `ENOENT`, `KP_ERRNO` combo.

File: src/main.c [hunk 13]
```diff
@@ -238,11 +223,36 @@ find_command(const char *command)
...
+		if (strncmp(argv[optind], "init", 5) != 0) {
```
Comment (paulfariello) [inline src/main.c]: I'd rather add this `if` after https://github.com/kickpass/kickpass/blob/72227445fd28c09e46a60cef9cf7ec8b924d97b3/src/main.c#L247-L251 so we don't try to init workspace and wait for an error.

**Review by paulfariello — COMMENTED** [review]
I like the `ws_fd`. There are minor things that could be enhance.
Thanks for the PR!

**Comment by paulfariello** [comment]
I just recall that I'd a feature/libkickpass branch. It creates a dedicated lib for kickpass with some cleaning and documentation of the core of kickpass. It may be a good idea to rebase workspace_fd on top of it.

**Comment by paulfariello** [comment]
i'll rebase your PR onto merged libkickpass.

**Comment by paulfariello** [comment]
I've rereviewed the PR the only remaining problem IMO is the handling of the `init` command.
It's really strange to cacth the ENOWORKSPACE error and still do the `kp_cmd_init`. Having `kp_init` returning an error code could mean the `struct kp_ctx` could be in a uninitialized state. I'd rather split the `kp_init` function and have the opening of the workspace in a dedicated `kp_open`.

**Comment by bapt** [comment]
I completly agree, I tried to make it the least intrusive way, to open the discussion, but I totally agree, we should first init then open

**Comment by paulfariello** [comment]
Ok so we can totally go with the `kp_open` function.

**Pull request merged**